### PR TITLE
WIP: Shore up doc comment handling

### PR DESCRIPTION
**Pull request by DJMcNab**

### CODE REVIEW

File: crates/ra_syntax/src/parser_impl/event.rs [hunk 1]
```diff
@@ -251,15 +251,26 @@ fn n_attached_trivias<'a>(
...
         CONST_DEF | TYPE_DEF | STRUCT_DEF | ENUM_DEF | FN_DEF | TRAIT_DEF | MODULE => {
             let mut res = 0;
+            // Whether we have stopped and are assumming that all later trivia is floating
+            let mut stopped = false;
             for (i, (kind, text)) in trivias.enumerate() {
                 match kind {
                     WHITESPACE => {
-                        if text.contains("\n\n") {
-                            break;
+                        if text.matches("\n").count() >= 2 && !stopped {
```
Comment (kjeremy) [inline crates/ra_syntax/src/parser_impl/event.rs]: FYI `matches` is disjoint so I think it will return 2 in the case of "\n\t\n".

Comment (DJMcNab) [reply crates/ra_syntax/src/parser_impl/event.rs]: That's why I wanted it - it seemed to make no sense that in
```rust
// This is a comment
 
/// This is also a comment
struct Foo{}
```
the entire file was is within the STRUCT, but in
```rust
// This is a comment

/// This is also a comment
struct Foo{}
```
the first comment is not

File: crates/ra_syntax/src/validation/comment.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    yellow::{
+        SyntaxError,
+        SyntaxErrorKind::*,
+    },
+    token_set::TokenSet,
+    Direction,
+    SyntaxKind::*
+};
+
+pub fn validate_comment(node: &ast::Comment, mut errors: &mut Vec<SyntaxError>) {
```
Comment (kjeremy) [inline crates/ra_syntax/src/validation/comment.rs]: Extra `mut`

**Comment by DJMcNab** [comment]
I'm going to restart this work with a different approach

**Pull request closed without merging**